File: cpp/stack/min_stack_leetcode_155.cpp
// using vector

/*
 
#include <bits/stdc++.h>
class MinStack {
  private: 
    std::vector<int> stack;
  public:
    MinStack(){}

    void pop(){
      if(!stack.empty()){
        stack.pop_back();
      } else {
        std::cout<<("stack is empty\n");
      }
    }

    void push(int item){
      stack.push_back(item);
    }

    int top(){
      if(!stack.empty()){
        return stack.back();
      } throw std::runtime_error("stack is empty");
    }

    int getMin(){
      if(!stack.empty()){
        return *min_element(stack.begin(), stack.end());
      } throw std::runtime_error("stack is empty");
    }
};

*/

// using stack
#include <iostream>
#include <stack>
#include <algorithm>

class MinStack{
  private:
    std::stack<int> stack;
    std::stack<int> minstack;

  public:
    MinStack(){}

    void push(int item){
      stack.push(item);

      if(minstack.empty() || item < minstack.top()){
        minstack.push(item);
      } else {
        minstack.push(minstack.top());
      }
    }

    void pop(){
      stack.pop();
      minstack.pop();
    }

    int top(){
      return stack.top();
    }

    int getMin(){
      return minstack.top(); 
    }
};

File: rust/stack/min_stack_leetcode_155.rs
struct MinStack {
    stack: Vec<i32>,
    minstack: Vec<i32>,
}

impl MinStack {
    fn new() -> Self {
        MinStack {
            stack: Vec::new(),
            minstack: Vec::new(),
        }
    }

    fn push(&mut self, val: i32) {
        self.stack.push(val);
        if self.minstack.is_empty() || val < *self.minstack.last().unwrap() {
            self.minstack.push(val);
        } else {
            self.minstack.push(*self.minstack.last().unwrap());
        }
    }

    fn pop(&mut self) {
        if self.stack.is_empty() {
            eprintln!("stack is empty");
        } else {
            self.stack.pop().unwrap();
            self.minstack.pop().unwrap();
        }
    }

    fn top(&self) -> i32 {
        *self.stack.last().unwrap()
    }

    fn get_min(&self) -> i32 {
        *self.minstack.last().unwrap()
    }
}

fn main() {
    let mut stack: MinStack = MinStack::new();
    stack.push(2);
    stack.push(4);
    stack.push(-2);
    println!("{}", stack.get_min());
    stack.pop();
    println!("{}", stack.top());
    println!("{}", stack.get_min());
}

